rfcs/ui-building-blocks.md

```diff
@@ -0,0 +1,472 @@
+However, it makes our components larger (reducing cache efficiency), and is a shift away from Plain Old Data,
+which makes the components more complex to work with as an end user.
+
+The only other obvious existing tools we have to do so are entity (or relevant component) cloning (very hard to implement, see [#1515](https://github.com/bevyengine/bevy/issues/1515)) or scenes.
+Scenes require a large amount of boilerplate to get working right now, including on every component, and are likely to be relatively expensive.
+
+## Unresolved questions
+
+1. Can / should we use trait queries to reduce the boilerplate involved in adding new style parameters?
+2. Does the `#[style_param]` magic work? Particularly with derive macros.
+3. Should we just use two generics in `add_style`, `propagate_style` and `maintain_style` and spare the macro magic?
+4. If we're using macro magic, do we care about the `Inner` trait type to avoid the querying footgun?
+5. Do we want to hold off on implementing this until after `min_relations` to get a substantially cleaner solution?
+
+## Future work
+
+1. Eventually, we can use the `Widget` and `Style` marker components to enforce appropriate `Entity` keys in both engine and user code using [kinded entities](https://github.com/bevyengine/bevy/issues/1634).
+2. Following the introduction of [archetype invariants](https://github.com/bevyengine/bevy/issues/1481),
+an archetype invariant should be added as part of `add_style`,
+which ensures that `Base` and `Final` variants of each component always coexist on `Widget` entities.
+3. Once we have fairly advanced [relations](https://github.com/bevyengine/bevy/pull/1627) features, we should look at using them in place of `Vec<Entity>` in `Styles`.
+This is not as trivial as it might appear: `Styles` very much wants an ordered list of entity references (rather than an unordered set or numbered priority list).
+4. As alternative to the mildly cursed `#[style_param]` solution, we could use [relations](https://github.com/bevyengine/bevy/pull/1627) to duplicate the data.
+The base version of the style parameter data will be a relation with a target of a unique dummy entity, whose identity is stored in a resource.
+Then, the final version of the data is a vanilla component (a relation with no target) of the same type,
+allowing end users to query for style parameter components directly to receive the current value after all styles have been applied.
+5. Nothing in the styling system limits its use to widgets alone. This opens up opportunities to use the same code and design patterns to game logic, meshes and so on. It hasn't been discussed above due to scope, but it's a powerful tool for end users.
+6. Finally, in order to fully move forward on the second rendition of `bevy_ui`, we also need to solve the following other UI focus areas:
+
+   1. Layout.
+   2. Wiring: call-backs, message passing (see #11).
+
+This RFC defines the underlying UI data structures, so should be settled first.
```

Comment:
I strongly disagree with this assertion. _Always write the usage code first_. [1]

When you write the usage code first, that interface will imply algorithms. Those algorithms will then imply the necessary data structures.

When you define data structures first, you end up contorting the algorithms and the user-focused API to fit those data structures. This is almost always less optimal in performance, maintainability, and quality.

The tricky part with a GUI is "what is the usage code"?

I have 2 answers for this.

1. There is the application code for querying the state of the GUI, processing GUI events, and updating or transforming the GUI at run-time.

2. There is the data definition of the GUI. If the intended way to develop a GUI is by writing application code, this is the same 1. However, if the intended way to develop a UI is through some type of non-code data format (whether that's XAML, HTML, XML, RON, etc) with a visual WYSIWYG editor, then this is _different_ from 1.

To that end, I would resolve the tracking issue for paper prototypes of the GUI [2] before this or any other GUI RFC (e.g. [3]) is written.  Each paper prototype should show a full vertical slice of GUI functionality. If you intend to have a data definition for the GUI, then each prototype should include that in the example. If the intent is that everything is defined in code, then the prototypes should show what not just snippets but full pseudocode for what layout, theming, composition, and data flow (data binding, polling, events, callbacks, etc) should look like in the context of that example.

It can be implemented in a straightforward manner. Start a Github repo and as each example is completed, put it through a PR. Reviewers throw stones at the examples _in context_. As each example is created, and new patterns are found or invalidated, the previous examples can be updated with additional PRs. At the end of the process, you'll have a Github repo containing each example with a description of it, annotated screenshots of the UI that you were trying to replicate, and pseudo-code of how you think the equivalent UI code would look for Bevy.

When you then write the RFCs, you _point at the Github repo of paper prototypes and say, "this is our proposal"_. The paper prototypes have already had significant peer review. They cover real use cases. They show full vertical slices. They can be compared to equivalent solutions written in other GUI frameworks, whether that's an entire example or just a particular feature ("here's how Elm UI would handle that").

Going through that process first has a significantly better chance of producing a usable, ergonomic API that can be implemented efficiently (with data structures chosen to fit the API afterward) _and_ it provides immediate feedback during the iterative approach of actually implementing the GUI once the RFCs (pointing to the prototypes!) are accepted: you just write code until each paper prototype isn't paper anymore!

From a scheduling perspective, you're going to go through that process anyway. The only question in my mind is how much time gets spent on these RFCs first.

Which leads to the next perspective -- architecture. These RFCs are not detailed enough for deep evaluation. If we wanted to debate the merits of this approach, the very first thing we'd need to do anyway is start mocking up larger examples and asking "ok, is this what it would look like? well what about this...". 

My last perspective is technical.

It is crucial to focus on usage code with realistic examples for 2 reasons.

1. It demonstrates that the proposed solution will, on paper, work.
2. It will focus the implementation to _exactly what is needed_.

In particular, I'm concerned by this line and other similar statements made with respect to the `relations` feature (not the subject of this RFC):

> Nothing in the styling system limits its use to widgets alone. This opens up opportunities to use the same code and design patterns to game logic, meshes and so on. It hasn't been discussed above due to scope, but it's a powerful tool for end users.

When developing any large codebase, _complexity_ is the enemy. Complexity and abstraction go hand in hand. Every single abstraction introduced in a codebase is a vector for technical debt. The best defense is to only introduce an abstraction when it has a clear need to exist and a clear scope of responsibilities. When you take a solution that was intended to solve one problem and apply it to another problem, it will often accumulate a little bit of complexity to meet the slightly different demands of the second problem. As you do this, for arbitrarily many problems, it becomes increasingly likely that the _one solution_ accumulates enough complexity that it is ultimately more error prone, less efficient, and less maintainable than it would have been to simply have _separate solutions for each problem_.

[1] https://caseymuratori.com/blog_0025
[2] https://github.com/bevyengine/bevy/issues/1974
[3] https://github.com/bevyengine/rfcs/pull/11

Reply:
Thanks for the feedback. Detail will come, as will end-to-end slices of complex examples.

Please follow https://github.com/bevyengine/bevy/issues/1974 for more updates along these lines, or feel free to start your own RFC outlining your vision for how UI should be done in Bevy.